File: elixir/2015/day03a/Program.exs
#!/usr/bin/env elixir

defmodule Program do
  def main(args) do
    case args do
      [filename] ->
        case process(filename) do
          {:ok, result} ->
            IO.puts("result = #{result}")
          {:error, reason} ->
            IO.puts(:stderr, "error while processing file `#{filename}': #{reason}")
            System.halt(1)
        end
      _ ->
        usage()
    end
  end

  defp usage do
    IO.puts(:stderr, "usage: Program.exs <input file>")
    System.halt(1)
  end

  defp process(filename) do
    case File.read(filename) do
      {:ok, content} ->
        {_, positions} =
          content
          |> String.graphemes()
          |> Enum.reduce({{0, 0}, MapSet.new()}, fn
            dir, {{santa_x, santa_y}, positions} -> 
              new_santa =
                case dir do
                  "^" -> {santa_x, santa_y + 1}
                  "v" -> {santa_x, santa_y - 1}
                  "<" -> {santa_x - 1, santa_y}
                  ">" -> {santa_x + 1, santa_y}
                  _ -> {santa_x, santa_y}
                end
              new_positions = MapSet.put(positions, new_santa)
              {new_santa, new_positions}
          end)
        {:ok, MapSet.size(positions)}
      {:error, reason} ->
        {:error, reason}
    end
  end
end

Program.main(System.argv())

File: erlang/2015/day03a/Program.erl
#!/usr/bin/env escript

main(Args) ->
  case Args of
    [Filename] ->
      case process(Filename) of
        {ok, Result} ->
          io:format("result = ~p~n", [Result]);
        {error, Reason} ->
          io:format(standard_error, "error while processing file `~s': ~p~n", [Filename, Reason]),
          halt(1)
      end;
    _ ->
      usage()
  end.

usage() ->
  io:format(standard_error, "usage: Program.erl <input file>~n", []),
  halt(1).

process(Filename) ->
  case file:read_file(Filename) of
    {ok, Chars} ->
      Content = unicode:characters_to_list(Chars),
      {_, Positions} = lists:foldl(
        fun
          (Ch, {{SantaX, SantaY}, Positions}) ->
            {NewSantaX, NewSantaY} =
              case Ch of
                $^ -> {SantaX, SantaY + 1};
                $v -> {SantaX, SantaY - 1};
                $< -> {SantaX - 1, SantaY};
                $> -> {SantaX + 1, SantaY};
                _ -> {SantaX, SantaY}
            end,
            NewPositions = sets:add_element({NewSantaX, NewSantaY}, Positions),
	    {{NewSantaX, NewSantaY}, NewPositions}
        end,
	{{0, 0}, sets:new()},
        Content),
      {ok, sets:size(Positions)};
    {error, Reason} ->
      {error, Reason}
  end.
